### DISCUSSION TITLE
antd5动态主题的最佳实践？

### DISCUSSION THREAD

**Discussion by Minf97** [ant-design/ant-design#46917] [Q&A] [answered]
我想请教一个问题：antd5如何实践**主题切换**？如：明亮|暗色模式

[官方文档描述](https://ant-design.antgroup.com/docs/react/customize-theme-cn#%E4%BD%BF%E7%94%A8%E9%A2%84%E8%AE%BE%E7%AE%97%E6%B3%95)
[掘金文档推荐](https://juejin.cn/post/7199832444836593725)

首先，在antd4之前，常见的做法是写CSS变量 `dark.less` 和 `light.less` ，通过切换CSS变量值，达到定制主题的目的。

现在antd5中，应该是希望通过 `CP组件` 的一套 `designToken` 实现切换。

---

那么问题来了，我实践中发现在我普通的 `div` 盒子中如文字颜色`color`，**依然需要CSS变量来控制切换主题，这样会增加复杂度，我们的预期应当是一套 `designToken` 即可，而不是 `designToken + CSS变量`** 

所以请问大佬们对于这一块是如何解决和实践的呢？

万分感谢&orz

**Comment by li-jia-nan** [comment]
没看懂，不是有 `theme.darkAlgorithm` 和 `theme.compactAlgorithm` 嘛？为啥需要手动控制 css 变量呢

**Reply by promiseLC** [reply]
他的意思是div里面文字color是样式文件里写的,改变了antd的theme,那非antd的样式该怎么处理

**Reply by Minf97** [reply]
antd组件的确可以通过`theme.darkAlgorithm`一键换肤，但在项目中不仅有antd的组件，**如普通的div盒子**，`theme.darkAlgorithm`无法支持普通的div盒子。

那么我想解决这个问题，是否需要用CSS变量了？

**Reply by arvinxx** [reply]
用 token 啊：
```ts
import { createStyles } from 'antd-style';

const useStyles = createStyles(({ token, css }) => ({
  // 支持 css object 的写法
  container: {
    backgroundColor: token.colorBgLayout,
    borderRadius: token.borderRadiusLG,
    maxWidth: 400,
    width: '100%',
    height: 180,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
    marginLeft: 'auto',
    marginRight: 'auto',
  },
  // 也支持通过 css 字符串模板获得和 普通 css 一致的书写体验
  card: css`
    color: ${token.colorTextTertiary};
    box-shadow: ${token.boxShadow};
    &:hover {
      color: ${token.colorTextSecondary};
      box-shadow: ${token.boxShadowSecondary};
    }

    padding: ${token.padding}px;
    border-radius: ${token.borderRadius}px;
    background: ${token.colorBgContainer};
    transition: all 100ms ${token.motionEaseInBack};

    margin-bottom: 8px;
    cursor: pointer;
  `,
}));

export default () => {
  // styles 对象在 useStyles 方法中默认会被缓存，所以不用担心 re-render 问题
  const { styles, cx, theme } = useStyles();

  return (
    // 使用 cx 可以组织 className
    <div className={cx('a-simple-create-style-demo-classname', styles.container)}>
      <div className={styles.card}>createStyles Demo</div>
      {/* theme 对象包含了所有的 token 与主题等信息 */}
      <div>当前主题模式：{theme.appearance}</div>
    </div>
  );
};
```

**Reply by Minf97** [reply]
感谢大佬！这样的确能**一套 `token` 解决所有问题**，但让我为难的是，我在平常中并**不习惯cssInJs**，我是使用scss；于是这个方案对我来说可能不是最优解，请问还有别的推荐解决方案吗？

**Reply by li-jia-nan** [reply]
这个不行吗？

```css
@media (prefers-color-scheme: dark) {

}
```

**Reply by Minf97** [reply]
既然是用antd5的`designToken`，我自然是希望一套`designToken`解决全部问题，不希望再使用CSS变量或媒体选择器了

**Reply by arvinxx** [reply]
> 感谢大佬！这样的确能**一套 `token` 解决所有问题**，但让我为难的是，我在平常中并**不习惯cssInJs**，我是使用scss；于是这个方案对我来说可能不是最优解，请问还有别的推荐解决方案吗？

还有个思路就是antd 开启 cssvar 模式。然后你的 scss 里写antd 的token 名

比如 `--ant-color-primary`，但这个方案我没验证过。 @MadCcc 看下是否可行

**Reply by Minf97** [reply]
在我的验证下是可行的，等我去翻翻所有可用的token里面 哪个token合适，，大佬能指个路吗，我不知道在哪里可以看到全部token

![image](https://github.com/ant-design/ant-design/assets/85478107/e8c330fc-c70c-4c2c-823c-574c5a586435)

![image](https://github.com/ant-design/ant-design/assets/85478107/85a1f243-5879-4acf-8b14-7defa701b291)

**Reply by li-jia-nan** [reply]
有个 `getDesignToken` 方法

**Reply by arvinxx** [reply]
https://ant-design.github.io/antd-token-previewer/~demos/docs-previewer-demo-previewer

**Comment by arvinxx** [comment, marked as answer]
https://ant-design.github.io/antd-style/zh-CN/guide/switch-theme

**Reply by Wxh16144** [reply]
https://ant-design.github.io/antd-style/zh-CN/guide/stylish 篇什么时候完善一下呀

**Reply by Minf97** [reply]
参考这篇文章中的业务代码，**不再使用div，转而用antd组件`Flex`或`Layout`替换**；**写文字时，不再使用text，转而用`Typography`替换**

上面这位大佬还提到用`CSS-IN-JS`实现**一套token解决全部主题切换的问题**，而无需再引入CSS变量或其他方法。

但我不习惯使用`CSS-IN-JS`，因此我选择尽量避免使用原生的html标签，而尽量都使用antd组件。

目前来看，我最常用的`div`和`text`都有了下位替代，而我一般不会使用到其他的原生标签，因此本问题结束~

```jsx
import { Typography } from 'antd';
const { Text } = Typography;

<Text className="自定义样式" />
```

**Reply by arvinxx** [reply]
这倒也是个思路

**Comment by arvinxx** [comment]
业务应用可以参考：https://github.com/lobehub/lobe-chat
